lib/Graylog.js

```diff
@@ -0,0 +1,427 @@
+
+		this.emit('chunk', header, data, i, chunkCount, server);
+
+		if (i < chunkCount - 1) {
+			client.send([header, data], server.port, server.host);
+
+			msgOffset += maxDataSize;
+		} else {
+			client.send([header, data], server.port, server.host, cb);
+		}
+	}
+};
+
+
+var count = 0;
+
+Graylog.prototype._tickDeflate = function () {
+	if (this._isDeflating || this._deflateQueue.isEmpty()) {
+		return;
+	}
+
+	this._isDeflating = true;
+
+	var that = this;
+	var msg = this._deflateQueue.getOne();
+
+	function done() {
+		that._isDeflating = false;
+		that._sendQueue.append(msg);
+
+		that._tickSend();
+		that._tickDeflate();
+	}
+
+	if (!this._alwaysDeflate && msg.buff.length <= this._bufferSize) {
+		process.nextTick(done);
+        return;
+    }
+
+    zlib.deflate(msg.buff, function (error, compressed) {
+        if (error) {
+            that.emit('warning', error);
+        } else {
+			that.compressed += 1;
+
+			if (that._alwaysDeflate || compressed.length < msg.buff.length) {
+				msg.buff = compressed;
+			}
+        }
+
+		done();
+    });
+};
+
+
+Graylog.prototype._tickSend = function () {
+	if (this._isSending) {
+		return;
+	}
+
+	if (this._sendQueue.isEmpty()) {
+		if (!this._isDeflating) {
+			this.emit('drain');
+		}
+		return;
+	}
+
+	this._isSending = true;
+
+	var that = this;
+	var msg = this._sendQueue.getOne();
+
+	function done(error) {
+		that._isSending = false;
+
+		if (error) {
+			that.emit('error', error);
+		}
+
+		that.sent += 1;
+		that._tickSend();
+	}
+
+	var buff = msg.buff;
+	msg.buff = null; // help GC a bit
+
+    if (buff.length <= this._bufferSize) {
+		// No need to chunk this message
+
+		var client = this._getClient();
+		var server = this._getServer();
+
+		this.emit('message', buff, server);
+
+		client.send(buff, 0, buff.length, server.port, server.host, done);
+		return;
+    }
+
+    // Generate a random ID (buffer)
+    crypto.randomBytes(8, function (error, id) {
+        if (error) {
+            return done(error);
+        }
+
+		that._sendChunked(id, buff, done);
+    });
+};
+
+
+Graylog.prototype._log = function log(level, short, full, fields, timestamp) {
+    if (!this._sendQueue) {
+		// destroyed
+		this.emit('warning', new Error('Trying to send on a closed client'));
+        return;
+    }
+
+    var message = {
+		buff: this._serialize(level, short, full, fields, timestamp),
+		next: null
+	};
+
+	if (!this._neverDeflate) {
+		this._deflateQueue.append(message);
+		this._tickDeflate();
+	} else {
+		this._sendQueue.append(message);
+		this._tickSend();
+	}
+};
+
+
+Graylog.prototype.close = function (cb) {
+	if (!cb) {
+		cb = function () {};
+	}
+
+    if (!this._isSending && !this._isDeflating) {
+        this.destroy();
+        return cb();
+    }
+
+    var that = this;
+
+    this.once('drain', function () {
+        that.destroy();
+		cb();
+    });
+};
+
+
+module.exports = Graylog;
+module.exports.graylog = Graylog; // deprecated
```

Comment:
Just to make sure, the rewrite does remains backward-compatible? If not, we might just remove this line flat-out.